File: OJ/leet004/python/leet004.py
# leet004

from typing import List


class Solution:
    def findMedianSortedArrays(self, A: List[int], B: List[int]) -> float:
        if len(A) > len(B): return self.findMedianSortedArrays(B, A)
        if len(A) == 0: 
            if len(B) % 2 == 0: return (B[len(B)//2] + B[len(B)//2-1]) / 2
            else: return B[len(B) // 2]

        half = (len(A) + len(B)) // 2
        lft, rht = 0, len(A)+1
        Alft, Blft = 0, 0
        while lft < rht:
            Alft = (lft + rht) // 2
            Blft = half - Alft
            # print(f'Alft = {Alft}, Blft = {Blft}, lft = {lft}, rht = {rht}')
            if Alft == 0:
                if A[Alft] < B[Blft-1]: lft = Alft + 1
                else: break
            elif Alft == len(A):
                if A[Alft-1] > B[Blft]: rht = Alft
                else: break 
            else:
                if A[Alft] < B[Blft-1]: lft = Alft + 1
                elif A[Alft-1] > B[Blft]: rht = Alft
                else: break
        if (len(A) + len(B)) % 2 == 0:
            if Alft == 0:
                mlft = B[Blft-1]
                mrht = A[0] if Blft == len(B) else min(A[0], B[Blft])
            elif Alft == len(A):
                mlft = A[Alft-1] if Blft == 0 else max(A[Alft-1], B[Blft-1])
                mrht = B[Blft]
            else:
                mlft = max(A[Alft-1], B[Blft-1])
                mrht = min(A[Alft], B[Blft])
            return (mlft + mrht) / 2
        else:
            if Alft == len(A): return B[Blft]
            else: return min(A[Alft], B[Blft])


# print(Solution().findMedianSortedArrays([1,2], [-1, 3]))
# print(Solution().findMedianSortedArrays([], [2, 3]))
print(Solution().findMedianSortedArrays([1], [2, 3]))
print(Solution().findMedianSortedArrays([], [2]))
print(Solution().findMedianSortedArrays([2], [2]))
print(Solution().findMedianSortedArrays([1,3], [2]))
print(Solution().findMedianSortedArrays([1,2], [3,4]))

File: OJ/leet004/leet004.cpp
/*
leet004


*/


#include "../utils/LeetCpp.utils.hpp"
using namespace std;

class Solution {
public:
    double findMedianSortedArrays(const vector<int>& A, const vector<int>& B) { 
        auto m = A.size(), n = B.size();
        auto K1 = (m + n + 1) / 2, K2 = (m + n + 2) / 2;
        return ( getKth(A, B, 0, m, 0, n, K1) +  getKth(A, B, 0, m, 0, n, K2) ) / 2.;
    }

    double getKth(const vector<int>& A, const vector<int>& B, int l1, int r1, int l2, int r2, int K) {
        int len1 = r1 - l1;
        int len2 = r2 - l2;
        if (len1 > len2) return getKth(B, A, l2, r2, l1, r1, K);
        if (len1 == 0) return B[l2+K-1];

        if (K == 1) return min(A[l1], B[l2]);

        int x = l1 + min(len1, K / 2) - 1;
        int y = l2 + min(len2, K / 2) - 1;

        if (A[x] > B[y]) return getKth(A, B, l1, r1, y+1, r2, K - (y - l2 + 1));
        else return getKth(A, B, x + 1, r1, l2, r2, K - (x - l1 + 1));
    }
};


int main(int argc, char const *argv[])
{
    Solution sol;   
    cout << sol.findMedianSortedArrays( {1, 3}, {2} ) << endl;
    cout << sol.findMedianSortedArrays( {1, 2}, {3,4} ) << endl;
    cout << sol.findMedianSortedArrays( {0, 0}, {0, 0} ) << endl;
    return 0;
}
